unit_tests/src/test_logical_locking.cpp

```diff
@@ -51,18 +76,141 @@ TEST_CASE("basic write lock")
+
+    // Open the test data object so that it will be write-locked
+    DataObjInp open_inp{};
+    auto cond_input = irods::experimental::make_key_value_proxy(open_inp.condInput);
+    const auto free_cond_input = irods::at_scope_exit{[&cond_input] { cond_input.clear(); }};
+    std::snprintf(open_inp.objPath, sizeof(open_inp.objPath), "%s", target_object.c_str());
+    cond_input[RESC_HIER_STR_KW] = opened_replica_resc;
+    open_inp.openFlags = O_WRONLY;
+
+    // Ensure that open successfully locked data object
+    const auto fd = rcDataObjOpen(&comm, &open_inp);
+    REQUIRE(fd > 2);
+    REQUIRE(INTERMEDIATE_REPLICA == ir::replica_status(comm, target_object, opened_replica_resc));
+    REQUIRE(WRITE_LOCKED == ir::replica_status(comm, target_object, other_replica_resc));
+
+    // This is done in an irods::at_scope_exit because catch will return immediately if
+    // a REQUIRE clause fails. This ensures that the data object is properly closed for
+    // cleanup purposes.
+    const auto close_object = irods::at_scope_exit{[&]
+    {
+        openedDataObjInp_t close_inp{};
+        close_inp.l1descInx = fd;
+        REQUIRE(rcDataObjClose(&comm, &close_inp) >= 0);
+
+        // Make sure everything unlocked properly
+        CHECK(GOOD_REPLICA == ir::replica_status(comm, target_object, opened_replica_resc));
+        CHECK(GOOD_REPLICA == ir::replica_status(comm, target_object, other_replica_resc));
+    }};
+
+    // Use a separate connection to run the tests from that used in the setup
+    irods::experimental::client_connection new_conn;
+    RcComm& new_comm = static_cast<RcComm&>(new_conn);
+
+    // Prepare input to rcDataObjOpen for the test cases
+    DataObjInp fail_open_inp{};
+    auto fail_cond_input = irods::experimental::make_key_value_proxy(fail_open_inp.condInput);
+    std::snprintf(fail_open_inp.objPath, sizeof(fail_open_inp.objPath), "%s", target_object.c_str());
+
+    //SECTION("open for read")
+    {
+        fail_open_inp.openFlags = O_RDONLY;
+
+        // open existing, currently opened replica
+        fail_cond_input[RESC_HIER_STR_KW] = opened_replica_resc;
+        REQUIRE(INTERMEDIATE_REPLICA_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+
+        // open existing, currently locked replica
+        fail_cond_input[RESC_HIER_STR_KW] = other_replica_resc;
+        REQUIRE(LOCKED_DATA_OBJECT_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+    }
+
+    //SECTION("open for write")
+    {
+        fail_open_inp.openFlags = O_WRONLY;
+
+        // open existing, currently opened replica
+        fail_cond_input[RESC_HIER_STR_KW] = opened_replica_resc;
+        REQUIRE(INTERMEDIATE_REPLICA_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+
+        // open existing, currently locked replica
+        fail_cond_input[RESC_HIER_STR_KW] = other_replica_resc;
+        REQUIRE(LOCKED_DATA_OBJECT_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+    }
+
+    //SECTION("open for create")
```

Comment:
this is commented out - not sure if that's intended.

Reply:
Added back the sections.